enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+While in the normal case this is discouraged, this could be used for short periods to test new configuration before
+rolling the updated configuration to all Control Plane Machines.
+
+Note that the OnDelete strategy is similar to the RollingUpdate strategy in that it needs to be able to add new
+Machines to the cluster to function. It can be thought of as a slower, more controlled RollingUpdate strategy where the
+user decides when each Control Plane Machine is replaced, and marks it for replacement by deleting it.
+Otherwise the replacement is identical to the RollingUpdate strategy.
+
+#### Protecting etcd quorum during update operations
+
+With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
+ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
+Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
+new instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneMachineSet
+
+As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
+`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
+the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
+instances. This will prevent the garbage collector from removing the Control Plane Machines when the
+`ControlPlaneMachineSet` is deleted.
+
+If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.
+
+#### Installing a ControlPlaneMachineSet within an existing cluster
+
+When adding a ControlPlaneMachineSet to a existing cluster, the end user will need to define the
+`ControlPlaneMachineSet` resource by copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
+field within the ControlPlaneMachineSet spec.
+
+To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
+that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.
+We will also validate that the failure domains align with those within the cluster already, this will prevent users
```

Comment:
How is this validation going to be done? Will you inspect cloud resources? Or will you limit the failure domains to just the those in use by active machines at the time that the CR is created?

Reply:
The intention is to do this via a ValidatingWebhook, looking at the active machines in the cluster at the time the CR is created, will need to identify whether this adds significant latency to that API create request or not, hopefully it doesn't.